File: src/GoogleQuestion.go
package main

import (
    "fmt"
    "math"
)

func main() {
    e := make([]int64, 100)
    for i := 0; i < 100; i++ {
        e[i] = 0
    }
    for i := getIterations(1000); i > 1; i-- {
        remainder := calculate(e, 0, i, 1)
        for j := 1; j < 100; j++ {
            remainder = calculate(e, j, i, remainder)
        }
    }

    result := e[0]
    totalPivot := getTotal(result)
    total := 0
    if isPrime(result) {
        fmt.Println(result)
        total = totalPivot
    }

    numbers := [10]int{}
    for _, value := range e[1 : ] {
        for i := 9; i >= 0; i-- {
            numbers[i] = int(value % 10)
            value /= 10
        }

        for _, number := range numbers {
            if 0 != total {
                total += number - int(result / int64(1e9))
                if total == totalPivot {
                    result = (result % int64(1e9)) * 10 + int64(number)
                    fmt.Println(result)
                    return
                }
            }

            result = (result % int64(1e9)) * 10 + int64(number)
            if 0 == total && isPrime(result) {
                fmt.Println(result)
                total = totalPivot
            }
        }
    }
}

func getIterations(digit int) int {
    result := 1
    remainder := 1.0 / math.E
    for digit >= 0 {
        for remainder < 1 {
            result += 1
            remainder *= float64(result)
        }
        for remainder >= 1 && digit >= 0 {
            remainder /= 10
            digit -= 1
        }
    }
    return result - 1
}

func calculate(e []int64, index int, number int, remainder int64) int64 {
    value := e[index] + remainder * int64(1e10)
    e[index] = value / int64(number)
    if (e[index] >= int64(1e10)) {
        e[index] -= int64(1e10)
        e[index - 1]++
    }
    return value % int64(number)
}

func getTotal(value int64) int {
    result := 0
    for 0 != value {
        result += int(value % 10)
        value /= 10
    }
    return result
}

func isPrime(value int64) bool {
    if 0 == (value & 1) {
        return false
    }

    border := int64(math.Sqrt(float64(value)))
    for i := int64(3); i < border; i += 2 {
        if 0 == (value % i) {
            return false
        }
    }
    return true
}

File: src/GoogleQuestion.py
import math
from typing import List

def main() -> None:
    e = [0 for _ in range(100)]
    for i in range(get_iterations(1000), 1, -1):
        remainder = calculate(e, 0, i, 1)
        for j in range(1, 100):
            remainder = calculate(e, j, i, remainder)

    result = e[0]
    total_pivot, total = get_total(result), 0
    if is_prime(result):
        print(result)
        total = total_pivot

    numbers = [0 for _ in range(10)]
    for value in e[1:]:
        for i in range(9, -1, -1):
            numbers[i] = value % 10
            value //= 10

        for number in numbers:
            if total:
                total += number - result // int(1e9)
                if total == total_pivot:
                    result = (result % int(1e9)) * 10 + number
                    print(result)
                    return

            result = (result % int(1e9)) * 10 + number
            if not total and is_prime(result):
                print(result)
                total = total_pivot

def get_iterations(digit: int) -> int:
    result, remainder = 1, 1.0 / math.e
    while digit >= 0:
        while remainder < 1:
            result += 1
            remainder *= result
        while remainder >= 1 and digit >= 0:
            remainder /= 10
            digit -= 1
    return result - 1

def calculate(e: List[int], index: int, number: int, remainder: int) -> int:
    value = e[index] + remainder * int(1e10)
    e[index] = value // number
    if e[index] >= int(1e10):
        e[index] -= int(1e10)
        e[index - 1] += 1
    return value % number

def get_total(value: int) -> int:
    result = 0
    while value:
        result += value % 10
        value //= 10
    return result

def is_prime(value: int) -> bool:
    if not (value & 1):
        return False

    border = int(math.sqrt(value))
    for i in range(3, border + 1, 2):
        if not value % i:
            return False
    return True

if __name__ == '__main__':
    main()
